### ISSUE TITLE
How to preserve client source IP (disable snat) at ingress gateway level (envoy) ?

### ISSUE THREAD

**Issue by jaxklag** [istio/istio#21527]
Hello,

**Problem description**
How to configure Istio CNI and Envoy proxy at ingress level with TPROXY mode ?

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
I would like to keep the client source IP when its TCP packet leave the istio ingress GW pod

**Version (include the output of istioctl version --remote and kubectl version and helm version if you used Helm)**

```
$ istioctl version --remote 
client version: 1.4.3
citadel version: 1.4.3
citadel version: 1.4.3
citadel version: 1.4.3
galley version: 1.4.3
galley version: 1.4.3
galley version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
pilot version: 1.4.3
pilot version: 1.4.3
pilot version: 1.4.3
policy version: 1.4.3
policy version: 1.4.3
policy version: 1.4.3
sidecar-injector version: 1.4.3
sidecar-injector version: 1.4.3
sidecar-injector version: 1.4.3
telemetry version: 1.4.3
telemetry version: 1.4.3
telemetry version: 1.4.3
data plane version: 1.4.3 (29 proxies), 1.4.0 (2 proxies)

$ kubectl version 
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.1", GitCommit:"d647ddbd755faf07169599a625faf302ffc34458", GitTreeState:"clean", BuildDate:"2019-10-02T17:01:15Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.1", GitCommit:"d647ddbd755faf07169599a625faf302ffc34458", GitTreeState:"clean", BuildDate:"2019-10-02T16:51:36Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

How was Istio installed?
istioctl

Istio is deployed with the following command (SDS & Istio CNI activated)

```
istioctl manifest generate --set profile=sds --set values.gateways.istio-egressgateway.enabled=false --set values.gateways.istio-ingressgateway.sds.enabled=true --set values.gateways.istio-ingressgateway.type=NodePort --set gateways.components.ingressGateway.k8s.hpaSpec.minReplicas=6 --set gateways.components.ingressGateway.k8s.hpaSpec.maxReplicas=8 --set values.global.mtls.enabled=false --set autoInjection.components.injector.k8s.replicaCount=3 --set configManagement.components.galley.k8s.replicaCount=3 --set trafficManagement.components.pilot.k8s.hpaSpec.minReplicas=3 --set policy.components.policy.k8s.hpaSpec.minReplicas=3 --set telemetry.components.telemetry.k8s.hpaSpec.minReplicas=3 -f istio-custom-resources.yaml > istio-sds-manifest-LoadBalancer8.yaml
istioctl manifest apply -f istio-sds-manifest-LoadBalancer8.yaml
```

Here the content of the istio-custom-resources.yaml  file
```yaml
$ cat istio-custom-resources.yaml 
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  cni:
    enabled: true
    components:
      cni:
        namespace: kube-system
  values:
    cni:
      excludeNamespaces:
       - default
       - istio-system
       - kube-node-lease
       - kube-public
       - kube-system
    mixer:
      telemetry:
        podAntiAffinityTermLabelSelector:
        - key: istio
          operator: In
          values: mixer
          topologyKey: failure-domain.beta.kubernetes.io/zone
          weight: 90
        - key: istio
          operator: In
          values: mixer
          topologyKey: kubernetes.io/hostname
          weight: 100
    pilot:
      podAntiAffinityTermLabelSelector:
      - key: istio
        operator: In
        values: pilot
        topologyKey: failure-domain.beta.kubernetes.io/zone
        weight: 90
      - key: istio
        operator: In
        values: pilot
        topologyKey: kubernetes.io/hostname
        weight: 100
    prometheus:
      podAntiAffinityTermLabelSelector:
      - key: app
        operator: In
        values: prometheus
        topologyKey: failure-domain.beta.kubernetes.io/zone
        weight: 90
      - key: app
        operator: In
        values: prometheus
        topologyKey: kubernetes.io/hostname
        weight: 100
    grafana:
      podAntiAffinityTermLabelSelector:
      - key: app
        operator: In
        values: grafana
        topologyKey: failure-domain.beta.kubernetes.io/zone
        weight: 90
      - key: app
        operator: In
        values: grafana
        topologyKey: kubernetes.io/hostname
        weight: 100
    gateways:
      istio-ingressgateway:
        ports:
        - name: status-port
          port: 15020
          targetPort: 15020
        - name: http2
          port: 80
          targetPort: 80
          nodePort: 30620
        - name: https
          port: 443
          nodePort: 31068
        - name: kiali
          port: 15029
          targetPort: 15029
        - name: prometheus
          port: 15030
          targetPort: 15030
        - name: grafana
          port: 15031
          targetPort: 15031
        - name: tracing
          port: 15032
          targetPort: 15032
        - name: tls
          port: 15443
          targetPort: 15443
  unvalidatedValues:
    cni:
      logLevel: info
    gateways:
      istio-ingressgateway:
        externalTrafficPolicy: Local
        podAntiAffinityTermLabelSelector:
        - key: app
          operator: In
          values: istio-ingressgateway
          topologyKey: failure-domain.beta.kubernetes.io/zone
          weight: 10
        - key: app
          operator: In
          values: istio-ingressgateway
          topologyKey: kubernetes.io/hostname
          weight: 20
        - key: app
          operator: In
          values: istio-ingressgateway
          topologyKey: accessed-from-internet
          weight: 100
  security:
    components:
      citadel:
        k8s:
          replicaCount: 3
```

Here is the infrastructure topology (very common).

Customer (IP 1.2.3.4) => LoadBalancer (HaProxy) => Kube Proxy (istio-ingressgateway) => Istio Ingress GW POD (envoy) => Customer POD (container + envoy sidecar)

All is working well into my k8s cluster and Istio Mesh. I'm using bookinfo application to test Istio.
Bookinfo application is reachable through HTTP and HTTPS.

My goal is to apply Istio AuthorizationPolicy based on source ipBlocks at the envoy proxy level of the customer POD. I want my customer to be able to filter the external trafic coming into their pods at the cusotmer POD level and not at the Ingress GW level.
To do that i need ingress GW Envoy proxy to preserve source IP (SNAT disable).

By default the souce IPs are natted by the envoy proxy of the Ingress GW using its own IP which is the IP of the ingress GW POD. So the packet coming into the envoy sidecar has the source IP of the Ingress GW POD and not the one of the real customer.

After many reads about the subject (tproxy mode) and certainly a misunderstanding from my side, i don't succeed in keeping the client source IP when the TCP packets goes outside the envoy proxy from the ingress gw pod.

I do not understand how to activate TPROXY mode on the Envoy proxy at the ingress GW  POD level.
Moreover, i think i need to modify Istio CNI DaemonSet to instruct it to use TPROXY mode to make iptables rules configured accordingly. How to do that with Istio CNI ?

Hope i've provided the needed information to get some help !
Thanks

**Comment by luisra51** [comment]
I think that you can change `externalTrafficPolicy: Cluster to externalTrafficPolicy: Local` in istio-ingressgateway service

**Comment by catman002** [comment]
@luisra51  can not work in istio1.4.5 or istio1.5

**Comment by luisra51** [comment]
@catman002 I try in 1.4.5 version and work

**Comment by jaxklag** [comment]
> 
> 
> I think that you can change `externalTrafficPolicy: Cluster to externalTrafficPolicy: Local` in istio-ingressgateway service

Already done.
The externalTrafficPolicy parameter set to Local at the service level allow to keep source IP when the packet leave the ipvs proxy. So the IP is preserved when coming into the ingress GW POD. But not when leaving the Ingress GW Pod which is what i want.

Here is my "schema" annotated which illustrate where my problem is.

Customer (IP 1.2.3.4) => LoadBalancer (HaProxy) (source ip 1.2.3.4 is preserved thanks to haproxy "usesrc clientip") => Kube Proxy (istio-ingressgateway) (source ip 1.2.3.4 is preserved thanks to service externalTrafficPolicy: Local parameter) => Istio Ingress GW POD (envoy) (How to preserve the source IP here when the packet leave the Envoy Proxy ?) =>  Customer POD (container + envoy sidecar) (Source IP is lost at this level because source ip is natted by Istio Ingress GW POD)

**Comment by jaxklag** [comment]
> 
> 
> @luisra51 can not work in istio1.4.5 or istio1.5

Are you telling me that it's impossible to preserve source IP after Istio ingress GW POD ?
How to do some ip based authorizationpolicies in that case ?

Thanks

**Comment by luisra51** [comment]
no,  @jaxklag 

- With the change of externalTrafficPolicy: Local I resolved to keep the source ip.
- The authorizationpolicies to restrict by ip did not work for a specific service. for the entire cluster yes

**Comment by zackb** [comment]
Hi @jaxklag, 
Did you ever find a solution to this? I'm facing the exact same issue. Using GCP network load balancers, externalTrafficPolicy Local, DaemonSets, I see the ingress GW gets the real user's IP but just the pod IP makes it to the service pods. Any help appreciated! Thanks

**Comment by jochenrichter** [comment]
Is there a solution for this issue? 
I have the same issue with an azure AKS cluster. externalTrafficPolicy Local is working fine without istio. But with istio inplace the request will be always natted with the ip of the ingressgateway pod

**Comment by FalconerTC** [comment]
Also curious on a resolution to this issue

**Comment by brentmjohnson** [comment]
Ran into this same issue, but with a bit of a different setup as i'm using apisix for the ingress gateway. Turns out i needed interceptionMode: TPROXY for this to work. I enabled this only for the ingress gateway pod(s) using the following annotation:

```
sidecar.istio.io/interceptionMode: TPROXY
```

see:

- https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig-InboundInterceptionMode
- https://istio.io/latest/docs/reference/config/annotations/

**Comment by rootsongjc** [comment]
I noticed that many of you have been inquiring about preserving client source IP in Istio. I recently wrote a blog post titled [Maintaining Traffic Transparency: Preserving Client Source IP in Istio](https://tetrate.io/blog/istio-source-ip-transparency/) that addresses this topic in detail. I hope you find it helpful in solving your source IP preservation challenges. Feel free to check it out and let me know if you have any questions or feedback.
